huggingface-gpt2/ort_addon/ort_supplement/src/transformers/gpt2_transformer.py

```diff
@@ -0,0 +1,314 @@
+                perm = node.attribute[0]
+                assert perm.name == 'perm'
+                perm = perm.ints
+                assert len(perm) == 2 and perm[0] == 1 and perm[1] == 0
+                transpose.append((get_node_index(model, node), weight))
+    for t in transpose:
+        node = model.graph.node[t[0]]
+        weight = numpy_helper.to_array(t[1])
+        assert len(weight.shape) == 2
+        weight = weight.transpose(perm)
+        new_weight = numpy_helper.from_array(weight, "%s_transposed" % t[1].name)
+        model.graph.initializer.extend([new_weight])
+        replace_input_arg(model, node.output[0], new_weight.name)
+
+    transpose.sort(reverse=True)
+    for t in transpose:
+        del model.graph.node[t[0]]
+
+    old_ws = []
+    old_graph_inputs=[]
+    for t in transpose:
+        if find_output_node(model, t[1].name) is None:
+            old_ws.append(find_weight_index(model, t[1].name))
+            old_graph_inputs.append(find_input_index(model, t[1].name))
+    old_ws.sort(reverse=True)
+    old_graph_inputs.sort(reverse=True)
+
+    for g_i in old_graph_inputs:
+        # print(model.graph.input[g_i].name)
+        del model.graph.input[g_i]
+
+    # print("clean up old weights")
+
+    for w_i in old_ws:
+        # print(model.graph.initializer[w_i].name)
+        del model.graph.initializer[w_i]
+
+def process_dropout(model):
+    dropouts = []
+    index = 0
+    for node in model.graph.node:
+        if node.op_type == 'Dropout':
+            new_dropout = model.graph.node.add()
+            new_dropout.op_type = 'TrainableDropout'
+            new_dropout.name = 'TrainableDropout_%d' % index
+            #make ratio node
+            ratio = np.asarray([node.attribute[0].f], dtype=np.float32)
+            # print(ratio.shape)
+            ratio_value = numpy_helper.from_array(ratio)
+            ratio_node = add_const(model, 'dropout_node_ratio_%d' % index, 'dropout_node_ratio_%d' % index, t_value=ratio_value)
+            # print (ratio_node)
+            new_dropout.input.extend([node.input[0], ratio_node.output[0]])
+            new_dropout.output.extend(node.output)
+            dropouts.append(get_node_index(model, node))
+            index += 1
+    dropouts.sort(reverse=True)
+    for d in dropouts:
+        del model.graph.node[d]
+
+def get_nodes_to_remove(input_id):
+    cast_node3 = find_input_node(model, input_id)
+    not_node3 = find_input_node(model, cast_node3.input[0])
+    if not_node3.op_type == "Not":
+        less_node = find_input_node(model, not_node3.input[0])
+    else:
+        assert not_node3.op_type == "Less"
+        less_node = not_node3
+    for less_input in less_node.input:
+        less_input_node = find_input_node(model, less_input)
+        if less_input_node and less_input_node.op_type == "Constant":
+            const_node = less_input_node
+            break
+    return [cast_node3, not_node3, less_node, const_node]
+
+def fix_split(model):
```

Comment:
I think the training is using opset 11, we can remove this step in post processing.